### PR TITLE
Add automatic memory limit handling

### PR DESCRIPTION
**Pull request by SuperQ**
Enable automatic detection of memory limits and configure GOMEMLIMIT to match.
* Also includes a flag to allow controlling the reserved ratio.

### CODE REVIEW

**Comment by SuperQ** [comment]
Depends on https://github.com/KimMachineGun/automemlimit/issues/13

File: cmd/prometheus/main.go [hunk 1]
```diff
@@ -262,6 +268,9 @@ func main() {
 	a.Flag("web.listen-address", "Address to listen on for UI, API, and telemetry.").
 		Default("0.0.0.0:9090").StringVar(&cfg.web.ListenAddress)
 
+	a.Flag("gomemlimit.ratio", "The ratio of detected memomry limit (container/system) used to set the GOMEMLIMIT").
```
Comment (bboreham) [inline cmd/prometheus/main.go:271]: “container/system” doesn’t resonate with me. 
Could you explain a bit what is meant?

I expected some note about needing space for memory-mapped IO.

Comment (SuperQ) [reply cmd/prometheus/main.go:271]: It first checks to see if it's in a cgroup and sets the limit if it finds a cgroup memory limit. If no cgroup limit is found it checks to total system memory and sets the limit based on that.

I can add some more of these details in the feature flag documentation page.

Comment (bboreham) [reply cmd/prometheus/main.go:271]: Ok but I’m asking on the “ratio” line.

**Comment by SuperQ** [comment]
@bboreham I added a documentation section to the feature flags docs.

File: cmd/prometheus/main.go [hunk 2]
```diff
@@ -262,6 +268,9 @@ func main() {
 	a.Flag("web.listen-address", "Address to listen on for UI, API, and telemetry.").
 		Default("0.0.0.0:9090").StringVar(&cfg.web.ListenAddress)
 
+	a.Flag("auto-gomemlimit.ratio", "The ratio of reserved GOMEMLIMIT memomry based on the detected container or system memory").
```
Comment (bboreham) [inline cmd/prometheus/main.go:271]:
```suggestion
	a.Flag("auto-gomemlimit.ratio", "The ratio of reserved GOMEMLIMIT memory to the detected maximum container or system memory").
```

**Review by bboreham — APPROVED** [review]
Code lgtm; now I think I understand what the docstring was trying to say I have suggested a rewording.

Also note CI failed last time.

File: cmd/prometheus/main.go (hunk 2 above)
Comment (SuperQ) [reply cmd/prometheus/main.go]: Good suggestion, thanks.

File: cmd/prometheus/main.go [hunk 3]
```diff
@@ -695,6 +709,21 @@ func main() {
...
+		_, err := memlimit.SetGoMemLimitWithOpts(
+			memlimit.WithRatio(cfg.memlimitRatio),
+			memlimit.WithProvider(
+				memlimit.ApplyFallback(
+					memlimit.FromCgroup,
+					memlimit.FromSystem,
+				),
+			),
+		)
+		if err != nil {
```
Comment (bwplotka) [inline cmd/prometheus/main.go:722]:
```suggestion
		if _, err := memlimit.SetGoMemLimitWithOpts(
			memlimit.WithRatio(cfg.memlimitRatio),
			memlimit.WithProvider(
				memlimit.ApplyFallback(
					memlimit.FromCgroup,
					memlimit.FromSystem,
				),
			),
		); err != nil {
```

**Review by bwplotka — APPROVED** [review]
LGTM, minor nit, but not a blocker.

File: cmd/prometheus/main.go [hunk 4]
```diff
@@ -262,6 +268,9 @@ func main() {
 	a.Flag("web.listen-address", "Address to listen on for UI, API, and telemetry.").
 		Default("0.0.0.0:9090").StringVar(&cfg.web.ListenAddress)
 
+	a.Flag("auto-gomemlimit.ratio", "The ratio of reserved GOMEMLIMIT memory to the detected maximum container or system memory").
```
Comment (machine424) [inline cmd/prometheus/main.go:271]: maybe we could also state here that it only takes effect when the feature `auto-gomemlimit` is enabled?

**Pull request merged**